pkg/controller/vsphere/actuator.go

```diff
@@ -61,6 +64,16 @@ func (a *Actuator) handleMachineError(machine *machinev1.Machine, err error, eve
 // Create creates a machine and is invoked by the machine controller.
 func (a *Actuator) Create(ctx context.Context, machine *machinev1.Machine) error {
 	klog.Infof("%s: actuator creating machine", machine.GetName())
+	// Ensure we're not reconciling a stale machine by checking our task-id.
+	// This is a workaround for a cache race condition.
+
+	if val, ok := a.taskIDCache[machine.Name]; ok {
+		if val != r.providerStatus.TaskRef {
+			klog.Errorf("%s: machine object missing expected provider task ID, requeue", machine.GetName())
+			return &machinecontroller.RequeueAfterError{RequeueAfter: requeueAfterSeconds * time.Second}
+		}
+	}
```

Comment:
If the patch fails to update the machine (taskRef will be empty), will this not result in an infinite loop? Constant re-queueing and never getting anywhere? 

What happens if the task ref is different to that in the cache? Is that a possibility? Could we not patch the machine with the TaskRef from this cache?

Reply:
No, we don't actually need to know the taskID for the subsequent Exists() calls to work on this provider.  In most cases, the actuator never goes down the Create() path twice, just in the cases where we double create, the instance cloning happens quite quickly to be discoverable by the machine's name.

Reply:
Ack yep, just double checked the code and can see that is the case

It does make me wonder how this double create can happen when we are setting the VM name and UUID to be predictable from the Machine. Surely as soon as a VM is created, we can find that VM by instance UUID?